crates/volta-core/src/tool/package/resolve.rs

```diff
@@ -0,0 +1,211 @@
+use volta_fail::{throw, Fallible, ResultExt};
+
+pub fn resolve(
+    name: &str,
+    matching: VersionSpec,
+    session: &mut Session,
+) -> Fallible<PackageDetails> {
+    let hooks = session.hooks()?.package();
+    match matching {
+        VersionSpec::Latest | VersionSpec::Lts => resolve_latest(name, hooks),
+        VersionSpec::Semver(requirement) => resolve_semver(name, requirement, hooks),
+        VersionSpec::Exact(version) => resolve_semver(name, VersionReq::exact(&version), hooks),
+    }
+}
+
+fn resolve_latest(name: &str, hooks: Option<&ToolHooks<Package>>) -> Fallible<PackageDetails> {
+    let package_index = match hooks {
+        Some(&ToolHooks {
+            latest: Some(ref hook),
+            ..
+        }) => {
+            debug!("Using packages.latest hook to determine package metadata URL");
+            let url = hook.resolve(&name)?;
+            resolve_package_metadata(name, &url)?.into()
+        }
+        _ => npm_view_query(name, "latest")?,
+    };
+
+    let latest = package_index.latest.clone();
+
+    let details_opt = match_package_details(package_index, |PackageDetails { version, .. }| {
+        &latest == version
+    });
+
+    match details_opt {
+        Some(details) => {
+            debug!(
+                "Found {} latest version ({}) from {}",
+                name, details.version, details.tarball_url
+            );
+            Ok(details)
+        }
+        None => Err(ErrorDetails::PackageVersionNotFound {
+            name: name.to_string(),
+            matching: "latest".into(),
+        }
+        .into()),
+    }
+}
+
+fn resolve_semver(
+    name: &str,
+    matching: VersionReq,
+    hooks: Option<&ToolHooks<Package>>,
+) -> Fallible<PackageDetails> {
+    let package_index = match hooks {
+        Some(&ToolHooks {
+            index: Some(ref hook),
+            ..
+        }) => {
+            debug!("Using packages.index hook to determine package metadata URL");
+            let url = hook.resolve(&name)?;
+            resolve_package_metadata(name, &url)?.into()
+        }
+        _ => npm_view_query(name, &matching.to_string())?,
+    };
+
+    let details_opt = match_package_details(package_index, |PackageDetails { version, .. }| {
+        matching.matches(&version)
+    });
+
+    match details_opt {
+        Some(details) => {
+            debug!(
+                "Found {}@{} matching requirement '{}' from {}",
+                name, details.version, matching, details.tarball_url
+            );
+            Ok(details)
+        }
+        None => Err(ErrorDetails::PackageVersionNotFound {
+            name: name.to_string(),
+            matching: matching.to_string(),
+        }
+        .into()),
+    }
+}
+
+/// Index of versions of a specific package.
+pub struct PackageIndex {
+    pub latest: Version,
+    pub entries: Vec<PackageDetails>,
+}
+
+// use the input predicate to match a package in the index
+fn match_package_details(
```

Comment:
The `match_` bit here naming-wise kind of threw me in the invocations above because I wanted to parse it as the start of a `match` block. My two thoughts are: `matching_package_details` and something like `details_for_package_matching`. (Not a deal-breaker *not* to do it, either.)

Reply:
As below, inlining the helper removes the naming issue.